test/parallel/test-buffer-concat.js

```diff
@@ -98,3 +99,25 @@ assert.deepStrictEqual(
 assert.deepStrictEqual(Buffer.concat([new Uint8Array([0x41, 0x42]),
                                       new Uint8Array([0x43, 0x44])]),
                        Buffer.from('ABCD'));
+        
+// Test buffer concatenation when total size will exceed 4 gigabytes.                   
+(function() {
+  // Don't run the test on 32-bit systems since we'll run out of memory. Or even
+  // if we make it to Buffer.concat somehow, we'll ERR_OUT_OF_RANGE trying to Buffer.allocUnsafe
+  // the destination.
+  if (2 ** 32 + 1 > kMaxLength) return;
+
+  let a = Buffer.alloc(2 ** 31);
+  let b = Buffer.alloc(2 ** 31);
+  let c = Buffer.alloc(1);
+  
+  a.fill(0);
+  b.fill(1);
+  c.fill(2);
+  
+  let destin = Buffer.concat([a, b, c]);
+  
+  assert.strictEqual(destin.subarray(0, 2 ** 31).compare(a), 0);
+  assert.strictEqual(destin.subarray(2 ** 31, 2 ** 32).compare(b), 0);
+  assert.equal(destin[2 ** 32], 2);
+})();
```

Comment:
```suggestion
if ((1 << 32) + 1 <= kMaxLength) {
  const buffer1 = Buffer.alloc(1 << 31, 0);
  const buffer2 = Buffer.alloc(1 << 31, 1);
  const buffer3 = Buffer.alloc(1, 2);

  const result = Buffer.concat([buffer1, buffer2, buffer3]);

  assert.deepStrictEqual(result.subarray(0, 1 << 31), buffer1);
  assert.deepStrictEqual(result.subarray(1 << 31, 1 << 32), buffer2);
  assert.deepStrictEqual(result.subarray(1 << 32), buffer3);
}
```

Reply:
1 << 32 wraps around back to 1 in JS.  Left shift implicitly converts its operands to 32-bit integers. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Left_shift

Reply:
Sure! However, the if statement is more in line with the rest of the file, as with the deepStrictEqual

Reply:
Sure! However, the if statement is more in line with the rest of the file, as with the deepStrictEqual

Reply:
Resolved by [3535bf9](https://github.com/nodejs/node/pull/55492/commits/3535bf923e6f20cab99abb60217f4b017a4197a1)